### ISSUE TITLE
Massive performance issues

### ISSUE THREAD

**Issue by derbolle** [wekan/wekan#2711]
## Issue

Add these issues to elsewhere:
- Snap: https://github.com/wekan/wekan-snap/issues

Other Wekan issues can be added here.

**Server Setup Information**:

* Did you test in newest Wekan?: yes (v3.41)

* Wekan version: v3.41
* Operating System: Linux: Debian 10 Buster
* Deployment Method(snap/docker/sandstorm/mongodb bundle/source): docker
* Http frontend if any (Caddy, Nginx, Apache, see config examples from Wekan GitHub wiki first): IIS reverse proxy 
![grafik](https://user-images.githubusercontent.com/53299502/64949039-96409200-d878-11e9-8be3-5544e3887d1d.png)

* Wekan only works on newest desktop Firefox/Chromium/Chrome/Edge/Chromium Edge and mobile Chrome. What webbrowser version are you using? Latest versions of chrome and Firefox

**Problem description**:

Hello there, 
since we updated wekan from an earlier version (I think it was 3.25) we have massive performance issues. 
- We have around 60 users 
- The db is around 0,1 gb small/large (it was 0,8 gb before the weekend somehow, maybe this is a display error)

![grafik](https://user-images.githubusercontent.com/53299502/64949759-33e89100-d87a-11e9-845f-ad10e46addae.png)

- The docker host-vm has 9GB of ram and 6 cores (Xeon Silver)

- Wekan app and db are the only containers which run on the server

- Traffic between wekan-app and wekan-db is massive(~30 minutes equals ~8gb of traffic between wekan-app and wekan-db. We have no large files as attachments etc. )
![grafik](https://user-images.githubusercontent.com/53299502/64950170-1667f700-d87b-11e9-838e-0c7aa333690c.png)

- load times are long as well (best case scenario ~20 seconds for a small board. Peak --> 20 minutes for a medium sized board (10 lists, ~100 cards))

Steps I've tried: 

- Assign more resources 
- Use different OS
- Use different Docker Version
- Completely redeploy to different server
- Use Chrome instead of Firefox

**Comment by hever** [comment]
Have you tried snap version? According to this wiki page https://github.com/wekan/wekan/wiki/Platforms I would say that snap version may be better. But i don't know just trying.

**Comment by xet7** [comment]
Newest Wekan is v3.42.

Load times affect only fist time Wekan board is loaded. After loading, using board is fast for me.

Wekan is single page app, so it is not meant that users keep pressing F5 or reloading webbrowser. You should navigate with links that are in in Wekan itself, like All Boards, Board, Card, etc.

After loading board, are there any errors at any of these?
- browser inspect console
- `docker logs wekan-app`
- `docker logs wekan-db`

Having some traffic to server with 60 users is not any bug.

**Comment by xet7** [comment]
In general, for large amount of users, options are:
- Use redis-oplog https://github.com/cult-of-coders/redis-oplog
- Use many Docker containers in front of Wekan, and scalable [AWS DocumentDB](https://aws.amazon.com/blogs/aws/new-amazon-documentdb-with-mongodb-compatibility-fast-scalable-and-highly-available/)
- Use some other scaling

**Comment by derbolle** [comment]
thanks, i will try and report back

**Comment by xet7** [comment]
Wekan has a lot of features:
- Real-time user interface, all changes to boards are visible to all users immediately.
- Rules for automatic actions.
- Outgoing Webhooks that can send board changes to Chat, NodeRED, etc immediately.
- Admin Panel has Global Webhooks that send can send all changes of all boards.
- There is possibility to have email notifications etc.

Yes, these do consume traffic.

**Comment by justinr1234** [comment]
Did you deploy the Wekan container with the mongo oplog url environment variable set?

**Comment by derbolle** [comment]
We have now deployed Wekan on a docker swarm:
- 1 management node, 2 CPUs, 4 GB Ram
- 2 worker nodes , 2 CPUs, $GB Ram each
- 8 Wekan App instances
- 1 mongo instance

we don't have a replicated mongodb so redis-oplog is out of the question for the moment(this would likely overcomplicate matters)

the aws documentation didn't really help for on premise use. I think a proper documentation how wekan can be used properly with docker on premise in production would help many other sysadmins, too. 
Furthermore I don't think the database is the bottleneck in our case. The wekan app itself seems to be the culprit here.

To the problem at hand: My first impression is that the user experience is a bit smoother with scaling out the wekan-app with docker swarm. BUT: After around one hour of uptime there is over 55 GB traffic between the apps and the db. This seems a bit high even if the performance has gotten a bit better. My first guess is that the "real time user interface" just loads everything from the db on a very fast schedule. The database is around 0,8 GB in size. This would mean that the whole database got queried over 60 times in the timeframe of as many minutes.  The resources wekan needs in general seem to be extraordinary high. 

![grafik](https://user-images.githubusercontent.com/53299502/65055843-33322680-d970-11e9-9f52-d78c2226f915.png)

**Comment by justinr1234** [comment]
@derbolle you need to check the environmental variables that are deployed with the Wekan app. You are likely missing the `MONGO_OPLOG_URL` variable which causes massive CPU performance problems.

This is almost guaranteed your problem.

**Comment by derbolle** [comment]
Hello @justinr1234 . The oplog URL is a suspicion I had before. Unfortunately I cannot find how it is supposed  to look like. The wekan documentation states that you need something like username:password@mongourl but the default docker container doesn't seem to have any mongo credentials. I am at a loss here

**Comment by justinr1234** [comment]
Post your dockerfiles here, and remove any sensitive credentials or information @derbolle

**Comment by derbolle** [comment]
Hello @justinr1234. Here is my docker-compose file with sensible infos like passwords and ldap removed

version: '3'

# Note: Do not add single quotes '' to variables. Having spaces still works without quotes where required.
#---------------------------------------------------------------------------------------------------------
# ==== CREATING USERS AND LOGGING IN TO WEKAN ====
# https://github.com/wekan/wekan/wiki/Adding-users
#---------------------------------------------------------------------------------------------------------
# ==== FORGOT PASSWORD ====
# https://github.com/wekan/wekan/wiki/Forgot-Password
#---------------------------------------------------------------------------------------------------------
# ==== Upgrading Wekan to new version =====
# 1) Stop Wekan:
#      docker-compose stop
# 2) Download new version:
#      docker-compose pull wekan
# 3) If you have more networks for VPN etc as described at bottom of
#    this config, download for them too:
#      docker-compose pull wekan2
# 4) Start Wekan:
#      docker-compose start
#----------------------------------------------------------------------------------
# ==== OPTIONAL: DEDICATED DOCKER USER ====
# 1) Optionally create a dedicated user for Wekan, for example:
#      sudo useradd -d /home/wekan -m -s /bin/bash wekan
# 2) Add this user to the docker group, then logout+login or reboot:
#      sudo usermod -aG docker wekan
# 3) Then login as user wekan.
# 4) Create this file /home/wekan/docker-compose.yml with your modifications.
#----------------------------------------------------------------------------------
# ==== RUN DOCKER AS SERVICE ====
# 1a) Running Docker as service, on Systemd like Debian 9, Ubuntu 16.04, CentOS 7:
#      sudo systemctl enable docker
#      sudo systemctl start docker
# 1b) Running Docker as service, on init.d like Debian 8, Ubuntu 14.04, CentOS 6:
#      sudo update-rc.d docker defaults
#      sudo service docker start
# ----------------------------------------------------------------------------------
# ==== USAGE OF THIS docker-compose.yml ====
# 1) For seeing does Wekan work, try this and check with your webbroser:
#      docker-compose up
# 2) Stop Wekan and start Wekan in background:
#     docker-compose stop
#     docker-compose up -d
# 3) See running Docker containers:
#     docker ps
# 4) Stop Docker containers:
#     docker-compose stop
# ----------------------------------------------------------------------------------
# ===== INSIDE DOCKER CONTAINERS, AND BACKUP/RESTORE ====
# https://github.com/wekan/wekan/wiki/Backup
# If really necessary, repair MongoDB: https://github.com/wekan/wekan-mongodb/issues/6#issuecomment-424004116
# 1) Going inside containers:
#    a) Wekan app, does not contain data
#         docker exec -it wekan-app bash
#    b) MongoDB, contains all data
#         docker exec -it wekan-db bash
# 2) Copying database to outside of container:
#      docker exec -it wekan-db bash
#      cd /data
#      mongodump
#      exit
#      docker cp wekan-db:/data/dump .
# 3) Restoring database
#      # 1) Stop wekan
#             docker stop wekan-app
#      # 2) Go inside database container
#             docker exec -it wekan-db bash
#      # 3) and data directory
#             cd /data
#      # 4) Remove previos dump
#             rm -rf dump
#      # 5) Exit db container
#             exit
#      # 6) Copy dump to inside docker container
#             docker cp dump wekan-db:/data/
#      # 7) Go inside database container
#             docker exec -it wekan-db bash
#      # 8) and data directory
#             cd /data
#      # 9) Restore
#             mongorestore --drop
#      # 10) Exit db container
#             exit
#      # 11) Start wekan
#             docker start wekan-app
#-------------------------------------------------------------------------

services:

  wekandb:
    #-------------------------------------------------------------------------------------
    # ==== MONGODB AND METEOR VERSION ====
    # a) For Wekan Meteor 1.8.x version at meteor-1.8 branch, use mongo 4.x
    #image: mongo:4.0.4
    # b) For Wekan Meteor 1.6.x version at master/devel/edge branches.
    # Only for Snap and Sandstorm while they are not upgraded yet to Meteor 1.8.x
    image: mongo:4.0.11-xenial
    deploy:
       placement: 
         constraints:
           - node.role == manager
    #-------------------------------------------------------------------------------------
    # container_name: wekan-db
    restart: always
    command: mongod --smallfiles --oplogSize 128
    networks:
      - wekan-tier
    expose:
      - 27017
    volumes:
      - wekan-db:/data/db
      - wekan-db-dump:/dump
  
  wekan:
    #-------------------------------------------------------------------------------------
    # ==== MONGODB AND METEOR VERSION ====
    # a) For Wekan Meteor 1.8.x version at meteor-1.8 branch,
    #    using https://quay.io/wekan/wekan automatic builds
    #image: quay.io/wekan/wekan:meteor-1.8
    # b) For Wekan Meteor 1.6.x version at master/devel/edge branches.
    # Only for Snap and Sandstorm while they are not upgraded yet to Meteor 1.8.x
    image: wekanteam/wekan:latest
    deploy:
      replicas: 8
    # c) Using specific Meteor 1.6.x version tag:
    # image: quay.io/wekan/wekan:v1.95
    # c) Using Docker Hub automatic builds https://hub.docker.com/r/wekanteam/wekan
    # image: wekanteam/wekan:meteor-1.8
    # image: wekanteam/wekan:v1.95
    #-------------------------------------------------------------------------------------
    # container_name: wekan-app
    restart: always
    networks:
      - wekan-tier
    #mem_limit: 3000m
    #-------------------------------------------------------------------------------------
    # ==== BUILD wekan-app DOCKER CONTAINER FROM SOURCE, if you uncomment these ====
    # ==== and use commands: docker-compose up -d --build
    #build:
    #  context: .
    #  dockerfile: Dockerfile
    #  args:
    #    - NODE_VERSION=${NODE_VERSION}
    #    - METEOR_RELEASE=${METEOR_RELEASE}
    #    - NPM_VERSION=${NPM_VERSION}
    #    - ARCHITECTURE=${ARCHITECTURE}
    #    - SRC_PATH=${SRC_PATH}
    #    - METEOR_EDGE=${METEOR_EDGE}
    #    - USE_EDGE=${USE_EDGE}
    #-------------------------------------------------------------------------------------
    ports:
      # Docker outsideport:insideport. Do not add anything extra here.
      # For example, if you want to have wekan on port 3001,
      # use 3001:8080 . Do not add any extra address etc here, that way it does not work.
      # remove port mapping if you use nginx reverse proxy, port 8080 is already exposed to wekan-tier network
     #redacted
    environment:
      - MONGO_URL=mongodb://wekandb:27017/wekan
      #---------------------------------------------------------------
      # ==== ROOT_URL SETTING ====
      # Change ROOT_URL to your real Wekan URL, for example:
      # If you have Caddy/Nginx/Apache providing SSL
      #  - https://example.com
      #  - https://boards.example.com
      # This can be problematic with avatars https://github.com/wekan/wekan/issues/1776
      #  - https://example.com/wekan
      # If without https, can be only wekan node, no need for Caddy/Nginx/Apache if you don't need them
      #  - http://example.com
      #  - http://boards.example.com
      #  - http://192.168.1.100    <=== using at local LAN
      - ROOT_URL=redacted
      #---------------------------------------------------------------
      # ==== EMAIL SETTINGS ====
      #redacted
      # ==== OPTIONAL: MONGO OPLOG SETTINGS =====
      # https://github.com/wekan/wekan-mongodb/issues/2#issuecomment-378343587
      # We've fixed our CPU usage problem today with an environment
      # change around Wekan. I wasn't aware during implementation
      # that if you're using more than 1 instance of Wekan
      # (or any MeteorJS based tool) you're supposed to set
      # MONGO_OPLOG_URL as an environment variable.
      # Without setting it, Meteor will perform a pull-and-diff
      # update of it's dataset. With it, Meteor will update from
      # the OPLOG. See here
      #   https://blog.meteor.com/tuning-meteor-mongo-livedata-for-scalability-13fe9deb8908
      # After setting
      # MONGO_OPLOG_URL=mongodb://<username>:<password>@<mongoDbURL>/local?authSource=admin&replicaSet=rsWekan
      # the CPU usage for all Wekan instances dropped to an average
      # of less than 10% with only occasional spikes to high usage
      # (I guess when someone is doing a lot of work)
      # - MONGO_OPLOG_URL=mongodb://<username>:<password>@<mongoDbURL>/local?authSource=admin&replicaSet=rsWekan
      #---------------------------------------------------------------
      # ==== OPTIONAL: KADIRA PERFORMANCE MONITORING FOR METEOR ====
      # https://github.com/edemaine/kadira-compose
      # https://github.com/meteor/meteor-apm-agent
      # https://blog.meteor.com/kadira-apm-is-now-open-source-490469ffc85f
      #- APM_OPTIONS_ENDPOINT=http://<kadira-ip>:11011
      #- APM_APP_ID=
      #- APM_APP_SECRET=
      #---------------------------------------------------------------
      # ==== OPTIONAL: LOGS AND STATS ====
      # https://github.com/wekan/wekan/wiki/Logs
      #
      # Daily export of Wekan changes as JSON to Logstash and ElasticSearch / Kibana (ELK)
      # https://github.com/wekan/wekan-logstash
      #
      # Statistics Python script for Wekan Dashboard
      # https://github.com/wekan/wekan-stats
      #
      # Console, file, and zulip logger on database changes https://github.com/wekan/wekan/pull/1010
      # with fix to replace console.log by winston logger https://github.com/wekan/wekan/pull/1033
      # but there could be bug https://github.com/wekan/wekan/issues/1094
      #
      # There is Feature Request: Logging date and time of all activity with summary reports,
      # and requesting reason for changing card to other column https://github.com/wekan/wekan/issues/1598
      #---------------------------------------------------------------
      # ==== WEKAN API AND EXPORT BOARD ====
      # Wekan Export Board works when WITH_API=true.
      # https://github.com/wekan/wekan/wiki/REST-API
      # https://github.com/wekan/wekan-gogs
      # If you disable Wekan API with false, Export Board does not work.
      - WITH_API=true
      #---------------------------------------------------------------
      # ==== PASSWORD BRUTE FORCE PROTECTION ====
      #https://atmospherejs.com/lucasantoniassi/accounts-lockout
      #Defaults below. Uncomment to change. wekan/server/accounts-lockout.js
      - ACCOUNTS_LOCKOUT_KNOWN_USERS_FAILURES_BEFORE=3
      - ACCOUNTS_LOCKOUT_KNOWN_USERS_PERIOD=60
      - ACCOUNTS_LOCKOUT_KNOWN_USERS_FAILURE_WINDOW=15
      - ACCOUNTS_LOCKOUT_UNKNOWN_USERS_FAILURES_BEFORE=3
      - ACCOUNTS_LOCKOUT_UNKNOWN_USERS_LOCKOUT_PERIOD=60
      - ACCOUNTS_LOCKOUT_UNKNOWN_USERS_FAILURE_WINDOW=15
      #---------------------------------------------------------------
      # ==== EMAIL NOTIFICATION TIMEOUT, ms =====
      # Defaut: 30000 ms = 30s
      #- EMAIL_NOTIFICATION_TIMEOUT=30000
      #-----------------------------------------------------------------
      # ==== CORS =====
      # CORS: Set Access-Control-Allow-Origin header.
      #- CORS=*
      #-----------------------------------------------------------------
      # ==== MATOMO INTEGRATION ====
      # Optional: Integration with Matomo https://matomo.org that is installed to your server
      # The address of the server where Matomo is hosted.
      #- MATOMO_ADDRESS=https://example.com/matomo
      # The value of the site ID given in Matomo server for Wekan
      #- MATOMO_SITE_ID=1
      # The option do not track which enables users to not be tracked by matomo
      #- MATOMO_DO_NOT_TRACK=true
      # The option that allows matomo to retrieve the username:
      #- MATOMO_WITH_USERNAME=true
      #-----------------------------------------------------------------
      # ==== BROWSER POLICY AND TRUSTED IFRAME URL ====
      # Enable browser policy and allow one trusted URL that can have iframe that has Wekan embedded inside.
      # Setting this to false is not recommended, it also disables all other browser policy protections
      # and allows all iframing etc. See wekan/server/policy.js
      - BROWSER_POLICY_ENABLED=true
      # When browser policy is enabled, HTML code at this Trusted URL can have iframe that embeds Wekan inside.
      #- TRUSTED_URL=https://intra.example.com
      #-----------------------------------------------------------------
      # ==== OUTGOING WEBHOOKS ====
      # What to send to Outgoing Webhook, or leave out. Example, that includes all that are default: cardId,listId,oldListId,boardId,comment,user,card,commentId .
      #- WEBHOOKS_ATTRIBUTES=cardId,listId,oldListId,boardId,comment,user,card,commentId
      #-----------------------------------------------------------------
      # ==== Debug OIDC OAuth2 etc ====
      #- DEBUG=true
      #-----------------------------------------------------------------
      # ==== OAUTH2 AZURE ====
      # https://github.com/wekan/wekan/wiki/Azure
      # 1) Register the application with Azure. Make sure you capture
      #    the application ID as well as generate a secret key.
      # 2) Configure the environment variables. This differs slightly
      #     by installation type, but make sure you have the following:
      #- OAUTH2_ENABLED=true
      # OAuth2 login style: popup or redirect.
      #- OAUTH2_LOGIN_STYLE=redirect
      # Application GUID captured during app registration:
      #- OAUTH2_CLIENT_ID=xxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxx
      # Secret key generated during app registration:
      #- OAUTH2_SECRET=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
      #- OAUTH2_SERVER_URL=https://login.microsoftonline.com/
      #- OAUTH2_AUTH_ENDPOINT=/oauth2/v2.0/authorize
      #- OAUTH2_USERINFO_ENDPOINT=https://graph.microsoft.com/oidc/userinfo
      #- OAUTH2_TOKEN_ENDPOINT=/oauth2/v2.0/token
      # The claim name you want to map to the unique ID field:
      #- OAUTH2_ID_MAP=email
      # The claim name you want to map to the username field:
      #- OAUTH2_USERNAME_MAP=email
      # The claim name you want to map to the full name field:
      #- OAUTH2_FULLNAME_MAP=name
      # Tthe claim name you want to map to the email field:
      #- OAUTH2_EMAIL_MAP=email
      #-----------------------------------------------------------------
      # ==== OAUTH2 KEYCLOAK ====
      # https://github.com/wekan/wekan/wiki/Keycloak  <== MAPPING INFO, REQUIRED
      #- OAUTH2_ENABLED=true
      # OAuth2 login style: popup or redirect.
      #- OAUTH2_LOGIN_STYLE=redirect
      #- OAUTH2_CLIENT_ID=<Keycloak create Client ID>
      #- OAUTH2_SERVER_URL=<Keycloak server name>/auth
      #- OAUTH2_AUTH_ENDPOINT=/realms/<keycloak realm>/protocol/openid-connect/auth
      #- OAUTH2_USERINFO_ENDPOINT=/realms/<keycloak realm>/protocol/openid-connect/userinfo
      #- OAUTH2_TOKEN_ENDPOINT=/realms/<keycloak realm>/protocol/openid-connect/token
      #- OAUTH2_SECRET=<keycloak client secret>
      #-----------------------------------------------------------------
      # ==== OAUTH2 DOORKEEPER ====
      # https://github.com/wekan/wekan/issues/1874
      # https://github.com/wekan/wekan/wiki/OAuth2
      # Enable the OAuth2 connection
      #- OAUTH2_ENABLED=true
      # OAuth2 docs: https://github.com/wekan/wekan/wiki/OAuth2
      # OAuth2 login style: popup or redirect.
      #- OAUTH2_LOGIN_STYLE=redirect
      # OAuth2 Client ID.
      #- OAUTH2_CLIENT_ID=abcde12345
      # OAuth2 Secret.
      #- OAUTH2_SECRET=54321abcde
      # OAuth2 Server URL.
      #- OAUTH2_SERVER_URL=https://chat.example.com
      # OAuth2 Authorization Endpoint.
      #- OAUTH2_AUTH_ENDPOINT=/oauth/authorize
      # OAuth2 Userinfo Endpoint.
      #- OAUTH2_USERINFO_ENDPOINT=/oauth/userinfo
      # OAuth2 Token Endpoint.
      #- OAUTH2_TOKEN_ENDPOINT=/oauth/token
      # OAuth2 ID Mapping
      #- OAUTH2_ID_MAP=
      # OAuth2 Username Mapping
      #- OAUTH2_USERNAME_MAP=
      # OAuth2 Fullname Mapping
      #- OAUTH2_FULLNAME_MAP=
      # OAuth2 Email Mapping
      #- OAUTH2_EMAIL_MAP=
      #-----------------------------------------------------------------
       # ==== LDAP: UNCOMMENT ALL TO ENABLE LDAP ====
     ##redacted
 

volumes:
  wekan-db:
  wekan-db-dump:

networks:
  wekan-tier:
   driver: overlay

**Comment by justinr1234** [comment]
@derbolle you need to uncomment the line for the OPLOG:

`MONGO_OPLOG_URL=mongodb://:@/local?authSource=admin&replicaSet=rsWekan`

Change it to:

`MONGO_OPLOG_URL=mongodb://wekandb:27017/local?authSource=admin&replicaSet=rsWekan

**Comment by derbolle** [comment]
Holy cow. If this works this would be awesome. I'll try immediately

**Comment by derbolle** [comment]
@justinr1234  sadly, this doesnt work. Wekan app containers fail with the following error. Could the reason be that we use mongo as a single instance or do i need to insert a environment variable to specify the oplog url in the wekan db configuration in docker-compose?

/build/programs/server/node_modules/fibers/future.js:313
                                                throw(ex);
                                                ^

MongoError: no primary found in replicaset or invalid replica set name
    at /build/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/topologies/replset.js:636:11
    at Server.<anonymous> (/build/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/topologies/replset.js:357:9)
    at Object.onceWrapper (events.js:315:30)
    at emitOne (events.js:116:13)
    at Server.emit (events.js:211:7)
    at /build/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/topologies/server.js:508:16
    at /build/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/pool.js:532:18
    at _combinedTickCallback (internal/process/next_tick.js:132:7)
    at process._tickCallback (internal/process/next_tick.js:181:9)

**Comment by justinr1234** [comment]
@derbolle try:

`MONGO_OPLOG_URL=mongodb://wekandb:27017/local`

**Comment by derbolle** [comment]
unfortunately still doesn't work. Thanks for all those hints by the way

/build/programs/server/node_modules/fibers/future.js:280
                                                throw(ex);
                                                ^

Error: $MONGO_OPLOG_URL must be set to the 'local' database of a Mongo replica set
    at OplogHandle._startTailing (packages/mongo/oplog_tailing.js:226:13)
    at new OplogHandle (packages/mongo/oplog_tailing.js:85:8)
    at new MongoConnection (packages/mongo/mongo_driver.js:225:25)
    at new MongoInternals.RemoteCollectionDriver (packages/mongo/remote_collection_driver.js:4:16)
    at Object.<anonymous> (packages/mongo/remote_collection_driver.js:38:10)
    at Object.defaultRemoteCollectionDriver (packages/underscore.js:784:19)
    at new Collection (packages/mongo/collection.js:97:40)
    at packages/idmontie_migrations.js:38:15
    at packages/idmontie_migrations.js:255:4
    at packages/idmontie_migrations.js:263:3
    at /build/programs/server/boot.js:419:36
    at Array.forEach (<anonymous>)
    at /build/programs/server/boot.js:228:19
    at /build/programs/server/boot.js:479:5
    at Function.run (/build/programs/server/profile.js:510:12)
    at /build/programs/server/boot.js:478:11

**Comment by justinr1234** [comment]
@derbolle try:

change `command: mongod --smallfiles --oplogSize 128` to `command: mongod --smallfiles --oplogSize 128 --replSet rsWekan`

Then for your oplog url: `MONGO_OPLOG_URL=mongodb://wekandb:27017/local?replicaSet=rsWekan`

**Comment by derbolle** [comment]
still no luck. we just use 1 mongo container. could that be the reason for the config not working?

/build/programs/server/node_modules/fibers/future.js:313
                                                throw(ex);
                                                ^

MongoError: no primary found in replicaset or invalid replica set name
    at /build/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/topologies/replset.js:636:11
    at Server.<anonymous> (/build/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/topologies/replset.js:357:9)
    at Object.onceWrapper (events.js:315:30)
    at emitOne (events.js:116:13)
    at Server.emit (events.js:211:7)
    at /build/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/topologies/server.js:508:16
    at /build/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/pool.js:532:18
    at _combinedTickCallback (internal/process/next_tick.js:132:7)
    at process._tickCallback (internal/process/next_tick.js:181:9)

**Comment by justinr1234** [comment]
@derbolle try:

`command: mongod --smallfiles --oplogSize 128 to command: mongod --smallfiles --oplogSize 128 --replSet rs0`

`MONGO_URL=mongodb://wekandb:27017/wekan`

`MONGO_OPLOG_URL=mongodb://wekandb:27017/local`

**Comment by derbolle** [comment]
Will try and answer asap.

**Comment by derbolle** [comment]
no luck 
Error: $MONGO_OPLOG_URL must be set to the 'local' database of a Mongo replica set
will try again in the morning
If I have success I will post my solution here for everyone to see. If anyone has any hint, I'll be grateful

**Comment by justinr1234** [comment]
I'm giving you the wrong commands, but it's in the right direction. You're missing the oplog.

**Comment by derbolle** [comment]
I have another suspicion: I think the command 
'mongod --smallfiles --oplogSize 128 to command: mongod --smallfiles --oplogSize 128 --replSet rs0'
will not create a replication set if there is an existing database without one specified (first run of container without --replset parameter)

I will try to remove the db container completely and create it anew with the --replset parameter. then I am going to do a mongorestore and see if the containers come up this time. 

I will report my results here but it'll have to wait until tonight/tomorrow

**Comment by derbolle** [comment]
I had the opportunity to test the following settings with a fresh install. (Docker swarm, single mongo instance, multiple wekan-app replicas)

as command in the docker compose (for wekandb)
'mongod --smallfiles --oplogSize 128 to command: mongod --smallfiles --oplogSize 128 --replSet rs0'

'MONGO_URL=mongodb://wekandb:27017/wekan'
'MONGO_OPLOG_URL=mongodb://wekandb:27017/local'

If I start the service with these settings the wekan-app containers shut down with the 
'Error: $MONGO_OPLOG_URL must be set to the 'local' database of a Mongo replica set'
error.

If I log into the wekan-db container and execute the mongo command 'rs.initiate()' the Wekan-app starts up seemingly fine

As soon as the container for wekan-db is recreated the wekan-app cannot access the database anymore because the db server is not a member of the replica set rs0 anymore. Every time the wekan-db container is recreated it gets assigned a new internal name and so is no member of the replica set anymore.

Unfortunately this means that the container cannot ever be recreated without manual work. 
 
Either this is a bug in the wekan docker-compose or wekan itself or I am doing something wrong here. If anything it should be mentioned in the docker-compose.yml that setting the Mongo Oplog url requires manually setting a replica set

Is this configuration working for anyone out there? (single database, mongo oplog url set)?

Any thoughts on this @xet7  and @justinr1234

**Comment by justinr1234** [comment]
@derbolle the command that runs when the container starts needs to be changed to run that initiate

**Comment by derbolle** [comment]
Yes, this is the problem. If I try to run multiple commands (rs initialization, then mongod start, we also tried the other way round), the wekan-db container does not work anymore. It seems to be that the replicaSet initialization has to be done manually (e.g. by a docker exec -it wekandbcontainer mongo rs.initiate(cconfig)) after creating the container. This makes automating wekan deployment with docker extremely difficult. Do you per chance know a better way of doing this? If so I would try to make it work and write up the documentation for everyone

**Comment by justinr1234** [comment]
@derbolle this might help, I didn’t read deeply https://github.com/docker-library/mongo/issues/246

It definitely can be done

**Comment by derbolle** [comment]
I will try and find a solution to this as soon as possible. 
With the workaround I described earlier we got the performance issues under control. In the current state wekan (with a single, containerized db instance) doesn't seem to be ready for production in environments larger than 20 users.  Likely we will move away from a dockerized mongodb to a dedicated db server/cluster. Nonetheless I will put in effort to make it work and post the solution in here, should I find it and contribute the changes to the docker-compose

**Comment by derbolle** [comment]
Hello there, 
our ongoing experimentation with wekan and docker have led us to a solution of this Problem. 
If you run Wekan on a Docker (swarm) with multiple instances of wekan-app  and one instance of wekan db you can initialize the replicaset while creating the container @xet7 : I will post the neccessary additions to the docker compose below. Please note that this notation is for running a docker swarm and you need to replace the deploy yml section with restart: on-failure because the syntax with docker and docker swarm is slightly different here. Also you need to make a few changes to the docker-compose in the wekan-app section and in the wekan-db section 

changes to wekan-app:
```
environment:
      - MONGO_URL=mongodb://wekandb:27017/wekan
      - MONGO_OPLOG_URL=mongodb://wekandb:27017/local


```
changes to wekan-db:

`
command: mongod --oplogSize 128  -replSet rs0
`

the additional container which should start just once:

` 
 wekandb-init-replica:

      image: mongo:latest

      deploy:
          restart_policy:
            condition: on-failure

      depends_on:
         - wekandb

      command: 'bash -c "for i in `seq 1 30`; do mongo wekandb/wekan --eval \"rs.initiate({ _id: ''rs0'', members: [ { _id: 0, host: ''localhost:27017'' } ]})\" && s=$$? && break || s=$$?; echo \"Tried $$i times. Waiting 5 secs...\"; sleep 5; done; (exit $$s)"'

      networks:
        - default
`

so far this did work splendidly - if you have any questions feel free to contact me

**Comment by Aversiste** [comment]
Hello @derbolle,

I am replying to your plea for other admins' experience with Wekan.
At my current work we have two instances: production and lab.

The production environment is a Debian 9 virtual machine with 16Go of RAM and 8 CPUs[1].
Wekan was installed using snap and is currently running version 3.84.
We have 50 active users on this instance and our biggest boards can take up to 15 seconds to load, which is too much but nothing like your 20 minutes.

The lab environment is a Debian 10 virtual machine with 4Go of RAM and 2 CPUs.
I experimented bypassing docker and snap with an ansible based install, using currently private roles to deploy meteor 1.9 and wekan 3.79.
It was a bit painful but it works.
I bootstrapped the database with a dump from production in order to have the same data.

I tried loading and measuring the same boards in both environment and the results are always very similar, averaging to the same numbers of seconds.
My next step is to configure mongodb's oplog as you did.

[1]: Absolutely overkill, our monitoring solution only shows 10% of memory consumption at best.

**Comment by derbolle** [comment]
hello @Aversiste ,

I absolutely cannot say anything about wekan instances which aren't deployed via docker. 

We are using a docker swarm with multiple (at the moment 3 i think)Wekan App instances and 3 docker workers. I absolutely recommend initializing the Replicaset for mongo even if you only have one DB instance like we have.  That did the trick for us
If you have any questions regarding this, I'll be glad to assist

**Comment by Aversiste** [comment]
This is what I did in my lab environment, with only one mongo instance running on the same server as wekan:

- First I adjusted my mongo configuration by creating and initializing a replicaset named `replica1` containing only localhost as a member ;
- I lowered the oplog size down to 128 ;
- I created a dedicated oplog user with `read` role on the `local` database ;
- I modified my systemd service template to include the environment variable `MONGO_OPLOG_URL` such as this: `MONGO_OPLOG_URL=mongodb://<username>:<password>@localhost/local?authSource=admin&replicaSet=replica1` ;
- Of course I restarted mongo and wekan.

Unfortunately I do not see any differences: the loading time for my boards is still around 10 seconds.

For me wekan is still slow, but thank you for your research anyway.
